NAME: **Ibeh Chimchetam Christabel**  
TRACK: **OIBSIP ( Data Analytics)**  
TITLE: **Cleaning Data**  
DATASET SOURCE: **Kaggle**

In [1]:
#Importing Libraries
import pandas as pd
import numpy as np

In [2]:
#Loading the dataset
df = pd.read_csv("C:/Users/Admin/Documents/My Data Sources/Titanic-Dataset.csv")

In [3]:
#initial inspection
print ("DATA QUALITY REPORT\n")
print(f"Shape:{df.shape}\n")
print(f"Duplicate rows:\n{df.duplicated().sum()}\n")
print(f"Dtypes:\n{df.dtypes}\n")
print(f"Nulls:\n{df.isnull().sum()}\n")
print(f"Summary statistics:\n{df.describe()}")

DATA QUALITY REPORT

Shape:(891, 12)

Duplicate rows:
0

Dtypes:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Nulls:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Summary statistics:
       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.0

**OBSERVATIONS:**  
The dataset contains 891 rows and 12 columns.  
**Nulls:** The age column has 177 empty cells, the cabin column has 687 empty cells, amd the Embarked column has only 2 empty cells.  
**Duplicates:** There are no duplicated rows.  
**Dtypes:** All columns are appropriately matched.  
**Value Range:** Age ranges from 0.42(children less than 1 year old) to 80. Fare ranges from 0.0 to 512.33.

In [4]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [5]:
#creating a copy of the datset
df_clean = df.copy()

In [6]:
#Handling Missing Data (Age)

#Extracting titles
df_clean['Title'] = df_clean['Name'].str.split(',').str[1].str.split('.').str[0].str.strip()

#Confirming output
print (df_clean['Title'].value_counts())

#Filling Null Age cells with median age
df_clean['Age'] = df_clean.groupby('Title')['Age'].transform(lambda x:x.fillna(x.median()))

#Confirming Correction Implementation
print (f"\nNull_Age: {df_clean['Age'].isnull().sum()}\n")


Title
Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Col               2
Mlle              2
Major             2
Ms                1
Mme               1
Don               1
Lady              1
Sir               1
Capt              1
the Countess      1
Jonkheer          1
Name: count, dtype: int64

Null_Age: 0



The **Age** column had 177 missing values. Rather than filling with an overall median, **age** was grouped by title and the missing values were filled with median age within each title group, giving a more reliable estimate.

In [7]:
#Handling Missing Data (Cabin)
df_clean['Has_Cabin'] = df_clean['Cabin'].apply(lambda x: 1 if pd.notnull (x)
                                                else 0)

#Confirming Correction Implementation
print(df_clean['Has_Cabin'].value_counts())

#Dropping Cabin column
df_clean.drop(columns=['Cabin'], inplace = True)

#Confirming Correction Implementation
df_clean.head()



Has_Cabin
0    687
1    204
Name: count, dtype: int64


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Title,Has_Cabin
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,Mr,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,Mrs,1
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,Miss,0
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,Mrs,1
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,Mr,0


The **Cabin** had a total of *687* missing values, i assumed these represented passengers without cabins. A new column named **Has_Cabin** was created, with all empty cells classified as **0** representing passengers *without* designated cabins and all others classified as *1*  representing passengers who were assigned cabins.  
I used binary so it corresponds with other variables in the dataset; to enhance uniformity.

In [8]:
#Handling Missing Data (Embarked)

#Filling Null cell with U (Unknown)
df_clean['Embarked'] = df_clean['Embarked'].fillna('U')

#Confirming Correction Implementation
print (df_clean['Embarked'].value_counts())

Embarked
S    644
C    168
Q     77
U      2
Name: count, dtype: int64


The **Embarked** column had *2* missing values. Rather than guessing and filling these missing values with a specific port, the empty cells were filled using *U* to represent *Unknown*, to enhance uniformity. 

In [9]:
#Duplicate rows removal
duplicate_count = df_clean.duplicated().sum()
print(f"Number of Duplicates Found: {duplicate_count}")

Number of Duplicates Found: 0


This dataset has no duplicates hence zero duplicates were removed.

In [10]:
#Standardization

#Checking for inconsistencies
print(f"{df_clean['Sex'].value_counts()}\n")
print(f"{df_clean['Title'].value_counts()}\n")
print(f"{df_clean['Embarked'].value_counts()}")

#Changing sex cells to title case
df_clean['Sex'] = df_clean['Sex'].apply(lambda x: x.title() if isinstance(x, str) else x)

#Confirming correction implementation
print(f"\nCorrected:\n{df_clean['Sex'].value_counts()}")


Sex
male      577
female    314
Name: count, dtype: int64

Title
Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Col               2
Mlle              2
Major             2
Ms                1
Mme               1
Don               1
Lady              1
Sir               1
Capt              1
the Countess      1
Jonkheer          1
Name: count, dtype: int64

Embarked
S    644
C    168
Q     77
U      2
Name: count, dtype: int64

Corrected:
Sex
Male      577
Female    314
Name: count, dtype: int64


There wasn't any major case of inconsistency within any column. The values in the sex column was changed to Title case just to correspond with values in other column within the dataset, so everything remains has a uniform appearance.

In [11]:
#Outlier detection (Fare)
#using Zscore
Mean_Fare = df_clean['Fare'].mean()
Std_Fare = df_clean['Fare'].std()

df_clean['Fare_Zscore'] = (df_clean['Fare'] - Mean_Fare)/ Std_Fare

Outliers = df_clean[df_clean['Fare_Zscore'].abs() > 3]
print(f"Number of Fare outliers: {len(Outliers)}")

Number of Fare outliers: 20


In [12]:
#Verification Check
Outliers.sort_values('Fare', ascending=False)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Title,Has_Cabin,Fare_Zscore
258,259,1,1,"Ward, Miss. Anna",Female,35.0,0,0,PC 17755,512.3292,C,Miss,0,9.661740
737,738,1,1,"Lesurer, Mr. Gustave J",Male,35.0,0,0,PC 17755,512.3292,C,Mr,1,9.661740
679,680,1,1,"Cardeza, Mr. Thomas Drake Martinez",Male,36.0,0,1,PC 17755,512.3292,C,Mr,1,9.661740
27,28,0,1,"Fortune, Mr. Charles Alexander",Male,19.0,3,2,19950,263.0000,S,Mr,1,4.644393
341,342,1,1,"Fortune, Miss. Alice Elizabeth",Female,24.0,3,2,19950,263.0000,S,Miss,1,4.644393
88,89,1,1,"Fortune, Miss. Mabel Helen",Female,23.0,3,2,19950,263.0000,S,Miss,1,4.644393
438,439,0,1,"Fortune, Mr. Mark",Male,64.0,1,4,19950,263.0000,S,Mr,1,4.644393
311,312,1,1,"Ryerson, Miss. Emily Borie",Female,18.0,2,2,PC 17608,262.3750,C,Miss,1,4.631815
742,743,1,1,"Ryerson, Miss. Susan Parker ""Suzette""",Female,21.0,2,2,PC 17608,262.3750,C,Miss,1,4.631815
118,119,0,1,"Baxter, Mr. Quigg Edmond",Male,24.0,0,1,PC 17558,247.5208,C,Mr,1,4.332899


In [13]:
#Checking total number of 1st class passengers
First_class_passengers = df_clean[df_clean['Pclass'] == 1]
print(f"Total First Class Passengers: {len(First_class_passengers)}")

#Getting Fare distribution among 1st class passengers, excluding outliers
First_class_fares = First_class_passengers[First_class_passengers['Fare_Zscore'].abs() <= 3]['Fare']

#Creating a maximum normal 1st class fare cap
fare_cap = First_class_fares.max()
print(f"Fare_Cap: {fare_cap:.2f}")

#Applying the Fare_cap to only outlier rows
df_clean.loc[df_clean['Fare_Zscore'].abs() > 3,'Fare'] = df_clean.loc[df_clean['Fare_Zscore'].abs() > 3,'Fare'].clip(upper=fare_cap)


Total First Class Passengers: 216
Fare_Cap: 164.87


In [14]:
#confirming capping was implemented
df_clean['Fare'].describe()

count    891.000000
mean      29.635367
std       35.532019
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      164.866700
Name: Fare, dtype: float64

**Observation:** The **Fare** column had a total of *20* outliers. A review of these rows showed that they all correspond with First class passengers, reflecting higher Fare cost. These outliers were *capped* to the maximum non-outlier fare observed within First class passengers specifically *(~164.87)*,this preserves realistic pricing, since First class fares are naturally cost more. 

In [15]:
#Outlier detection (Age)
#using Zscore
Mean_Age = df_clean['Age'].mean()
Std_Age = df_clean['Age'].std()

df_clean['Age_Zscore'] = (df_clean['Age'] - Mean_Age)/ Std_Age

Age_outliers = df_clean[df_clean['Age_Zscore'].abs() > 3]
print(f"Number of Age outliers: {len(Age_outliers)}")

Number of Age outliers: 7


In [16]:
#Verification Check
Age_outliers.sort_values('Age', ascending=False)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Title,Has_Cabin,Fare_Zscore,Age_Zscore
630,631,1,1,"Barkworth, Mr. Algernon Henry Wilson",Male,80.0,0,0,27042,30.0000,S,Mr,1,-0.044356,3.815196
851,852,0,3,"Svensson, Mr. Johan",Male,74.0,0,0,347060,7.7750,S,Mr,0,-0.491598,3.362888
96,97,0,1,"Goldschmidt, Mr. George B",Male,71.0,0,0,PC 17754,34.6542,C,Mr,1,0.049302,3.136735
493,494,0,1,"Artagaveytia, Mr. Ramon",Male,71.0,0,0,PC 17609,49.5042,C,Mr,0,0.348134,3.136735
116,117,0,3,"Connors, Mr. Patrick",Male,70.5,0,0,370369,7.7500,Q,Mr,0,-0.492101,3.099043
672,673,0,2,"Mitchell, Mr. Henry Michael",Male,70.0,0,0,C.A. 24580,10.5000,S,Mr,0,-0.436762,3.061350
745,746,0,1,"Crosby, Capt. Edward Gifford",Male,70.0,1,1,WE/P 5735,71.0000,S,Capt,1,0.780703,3.061350


**Observation:** Also using the Z-score method, *7* outliers were identified in the **Age** column. These rows corresponded to elderly **passengers** . These outliers were *retained* because capping or removing them would misrepresent actual passenger information.

In [17]:
#Data type correction
print(df_clean.dtypes)

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
Title           object
Has_Cabin        int64
Fare_Zscore    float64
Age_Zscore     float64
dtype: object


**Observation:** As shown in the initial data quality report carried out on the dataset, all columns are already appropriately typed.

In [18]:
#Before and After Summary Table
summary = pd.DataFrame({
    'Metric': ['Missing Age', 'Missing Cabin', 'Missing Embarked', 'Duplicate Rows', 'Total Rows'],
    'Before': [
        df['Age'].isnull().sum(),
        df['Cabin'].isnull().sum(),
        df['Embarked'].isnull().sum(),
        df.duplicated().sum(),
        len(df)
    ],
    'After': [
        df_clean['Age'].isnull().sum(),
        'N/A (converted to Has_Cabin)',
        df_clean['Embarked'].isnull().sum(),
        df_clean.duplicated().sum(),
        len(df_clean)
    ]
})

summary


,Metric,Before,After
0,Missing Age,177,0
1,Missing Cabin,687,N/A (converted to Has_Cabin)
2,Missing Embarked,2,0
3,Duplicate Rows,0,0
4,Total Rows,891,891


In [19]:
outlier_summary = pd.DataFrame({
    'Column': ['Fare', 'Age'],
    'Outliers Found (Z-score)': [20, 7],
    'Action Taken': ['Capped to 1st class max fare', 'Retained — valid elderly passengers']
})

outlier_summary

,Column,Outliers Found (Z-score),Action Taken
0,Fare,20,Capped to 1st class max fare
1,Age,7,Retained — valid elderly passengers


In [20]:
#Saving cleaned dataset to a new CSV file 
df_clean.to_csv('Titanic_Cleaned.csv', index=False)